src/GlobalState.js

```diff
@@ -0,0 +1,155 @@
+
+const currentMonth = getCurrentMonth();
+
+const initialState = {
+  expenses: [],
+  selectedMonthExpenses: [],
+  currentMonth: currentMonth,
+  selectedMonth: currentMonth,
+  lastAddedExpense: {},
+  expenseToDelete: {},
+  expenseToEdit: {},
+  isConfirmMonthModalOpen: false,
+  isConfirmDeleteModalOpen: false,
+  isEditExpenseModalOpen: false,
+  didErrorOccur: false,
+  isAlertOpen: false,
+  userAction: '',
+};
+
+export const GlobalContext = createContext(initialState);
+
+export const GlobalProvider = ({ children }) => {
+  const [globalState, dispatch] = useReducer(appReducer, initialState);
+
+  const { expenses, selectedMonth, currentMonth, selectedMonthExpenses } = globalState;
+
+  // Listen for changes to any doc in "expenses" collection and update expenses in app
+  useEffect(() => {
+    try {
+      const q = query(collection(db, 'expenses'));
+
+      onSnapshot(q, (querySnapshot) => {
+        const expenses = [];
+        querySnapshot.forEach(doc => {
+          let expense = { ...doc.data(), id: doc.id };
+          expenses.push(expense);
+        });
+        dispatch({ type: 'RESTORE_EXPENSES', payload: expenses });
+      });
+    } catch (err) {
+      console.log(err);
+      alert('Error loading expenses. Please try again later.');
+    }
+  }, []);
+
+  // Filter for expenses that belong in the selected month
+  useEffect(() => {
+    const filteredExpenses = expenses
+      .filter(expense => {
+        const expenseMonth = getUKFormattedDate(new Date(expense.date), { year: 'numeric', month: 'long' });
+        return selectedMonth === expenseMonth;
+      })
+      .sort((expenseA, expenseB) => new Date(expenseB.date) - new Date(expenseA.date));
+
+    dispatch({ type: 'UPDATE_SELECTED_MONTH_EXPENSES', payload: filteredExpenses });
+  }, [expenses, selectedMonth]);
+
+  useEffect(() => {
+    if (selectedMonthExpenses.length === 0) {
+      changeMonthView(currentMonth);
+    }
+  }, [selectedMonthExpenses, currentMonth]);
+
+  const globalFunctions = {
+    addExpense,
+    editExpense,
+    deleteExpense,
+    changeMonthView,
+    openConfirmDeleteModal,
+    closeConfirmDeleteModal,
+    openEditExpenseModal,
+    closeEditExpenseModal,
+    closeConfirmMonthModal,
+    closeAlert,
+  };
+
+  async function addExpense(expense) {
+    try {
+      const expensesRef = collection(db, 'expenses');
+      await addDoc(expensesRef, expense);
+      dispatch({ type: 'ADD_EXPENSE_SUCCESS', payload: expense });
+    } catch (err) {
+      console.log(err);
+      dispatch({ type: 'ADD_EXPENSE_FAIL' });
+    }
+  }
+
+  async function editExpense(id, newExpenseData) {
+    let didErrorOccur = false;
+
+    try {
+      const expenseRef = doc(db, 'expenses', id);
+      await updateDoc(expenseRef, newExpenseData);
+    } catch (err) {
+      console.log(err);
+      didErrorOccur = true;
+    }
+
+    dispatch({ type: 'EDIT_EXPENSE', payload: didErrorOccur });
```

Comment:
question: should this not be inside the catch block?

Reply:
I don't think so. We want to dispatch the action whether the catch block runs or not, and we also want to allow the value of `didErrorOccur` to be either `true` or `false` so that we trigger the correct alert.

Reply:
ah ok understand... the alternative would be then something like :
`
try {
... ;
dispatch({ type: ... , payload: didErrorOccur = false }) 
} catch (error) {
... ;
dispatch({ type: ... , payload: didErrorOccur = true })
} `

I get it now... either way it works!

Reply:
Ah cool, I didn't know you could set the value of a property like that `{ ..., payload: didErrorOccur = true }`

Is there a name for this pattern?

Reply:
sorry man. I was just mocking the alternative way... not really writing exact code. no new interesting pattern 😄

Reply:
Hmm, I tried it out in the console, and it works. It's almost like a way to apply a temporary variable name to a property value to make it clearer what it is, but I'm not sure if it's actually something people use in practice. 

`{ propertyName: someUndeclaredVariable = value }`

@motine Do you know?

Reply:
the result of an assignment in javascript is the assigned value. so it works as intended. using assignments as parameters however is often frowned upon, because one might overlook the assignment when searching for it. i would personally try to rename `propertyName` if possible (should be in this case). if this is not possible, then i would extract the variable  assignment to the line above and then pass the variable. this way, you have the best of both (descriptive, no hidden assignment).